scylla/src/observability/metrics.rs

```diff
@@ -33,13 +33,181 @@ pub struct Snapshot {
     pub percentile_99_9: u64,
 }
 
+/// The interval in seconds for which the rate is calculated.
+const INTERVAL: u64 = 5;
+
+/// An implementation of an Exponentially Weighted Moving Average (EWMA).
+#[derive(Debug)]
+struct ExponentiallyWeightedMovingAverage {
+    /// Smoothing factor, a value between 0 and 1.
+    alpha: f64,
+    /// Atomic counter to keep track of the number of requests. \
+    /// Indicates the number of requests that have not been accounted for EWMA yet.
+    uncounted: AtomicU64,
+    /// To check if the EWMA has been initialized.
+    is_initialized: Mutex<bool>,
+    ///  Atomic value representing the current rate of requests per second. \
+    ///  AtomicU64 is used to store a floating point number as a bit representation.
+    rate: AtomicU64,
+}
+
+impl ExponentiallyWeightedMovingAverage {
+    fn new(alpha: f64) -> Self {
+        Self {
+            alpha,
+            uncounted: AtomicU64::new(0),
+            is_initialized: Mutex::new(false),
+            rate: AtomicU64::new(0),
+        }
+    }
+
+    /// Returns the current `rate` as a floating point number.
+    fn rate(&self) -> f64 {
+        f64::from_bits(self.rate.load(Ordering::Acquire))
+    }
+
+    /// Increments the `uncounted` requests counter. \
+    /// Should be called every time a new request is made.
+    fn update(&self) {
+        self.uncounted.fetch_add(1, ORDER_TYPE);
+    }
+
+    /// Updates the `rate` based on the current number of requests. \
+    /// Should be called every time the interval has passed.
+    ///
+    /// The rate is updated using the formula: \
+    /// `rate = rate + alpha * (instant_rate - rate)` \
+    /// where `instant_rate` is the number of requests in the last interval.
+    ///
+    /// The first time this function is called, the `rate` is set to the `instant_rate`.
+    fn tick(&self) {
+        let count = self.uncounted.swap(0, ORDER_TYPE);
+        let instant_rate = count as f64 / INTERVAL as f64;
+
+        let mut is_initialized = self.is_initialized.lock().unwrap();
+        if *is_initialized {
+            let rate = f64::from_bits(self.rate.load(Ordering::Acquire));
+            self.rate.store(
+                f64::to_bits(rate + self.alpha * (instant_rate - rate)),
+                Ordering::Release,
+            );
+        } else {
+            self.rate
+                .store(f64::to_bits(instant_rate), Ordering::Release);
+            *is_initialized = true;
+        }
```

Comment:
I think this code is still not fully correct. Previously I suggested locking the mutex before the `count` is read. Why?
Consider the scenario, where driver has a constant rate of queries, and was just started, and we got 2 concurrent calls to this method.
Following execution is possible:
- Thread 1 reads and zeros count
- Thread 2 reads 0
- Thread 2 gets mutex, `is_initialized` is false, so goes to the second branch
- It stores `0` rate
- Thread 1 gets mutex, stores new rate

Now the rate will slowly increase, instead of being correct from the beginning.
Please move the lock before count read, I see no downsides to it.